src/arnoldi.jl

```diff
@@ -105,6 +97,98 @@ for (iter, Ar) in enumerate(Arnoldi(K, Terminator(1e-9, n)))
     @assert abs(norm(K.A*Ar.V-Ar.V*Ar.H) - norm(Ar.r)) < sqrt(eps())
 end
 
+# Implicitly restarted Arnoldi (Algorithm 3.8; Sorensen, 1992)
+# XXX the signature should really be chained: T, Alg<:Factorizer{T}
+type ImplicitlyRestarted{Alg<:Factorizer} <: Factorizer
+    P :: Alg
+    k :: Int #Maximum size of Arnoldi factorization
+    p :: Int #Number of extra steps
+    getshifts :: Function
+end
+
+function ImplicitlyRestarted{Alg<:Factorizer}(P::Alg, k::Int, p::Int, getshifts::Function=Shifts)
+    ImplicitlyRestarted{Alg}(P, k, p, getshifts)
+end
+
+#Explicitly restarted
+type Restarted{Alg<:Factorizer} <: Factorizer
+    P :: Alg
+    k :: Int #Maximum size of Arnoldi factorization
+end
+
+start(P::Union(ImplicitlyRestarted,Restarted)) = start(P.P)
+
+function next{T}(P::Restarted{Arnoldi{T}}, F::ArnoldiFact{T})
+    F, _ = next(P.P, F)
+    if size(F.H,1) == P.k
+        n = size(F.r, 1)
+        V = zeros(T,n,0)
+        H = zeros(T,0,0)
+        F = ArnoldiFact{T}(V, H, F.r)
+    end
+    F, F
+end
+
+function next{T}(P::ImplicitlyRestarted{Arnoldi{T}}, F::ArnoldiFact{T})
+    F, _ = next(P.P, F)
+    if size(F.H,1)==P.k+P.p #Deflate down to size k again
+        k, p = P.k, P.p
+        H, V, r = F.H, F.V, F.r
+
+        u = P.getshifts(H, P.p)
+
+        Q = I
+        for j=1:p
+            Qj, Rj = qr(H-u[j]*I)
+            H = Qj'H*Qj #TODO replace with bulge-chasing
+            Q = Q*Qj
+        end
+
+        VQ = V*Q
```

Comment:
I think this should be done in-place, because V is most memory-consuming. If you do restarts because of memory limitations, then this procedure would require you to restart twice as early. One way to go is use the LU-trick: V * Q = (V * L) * U. Since L & U are upper triangular, this can be done in-place.